tests/Agent/IntegrationTests/IntegrationTestHelpers/RandomPortGenerator.cs

```diff
@@ -16,22 +18,29 @@ public static class RandomPortGenerator
         private const int maxPortID = 60000;
         private const int portPoolSize = maxPortID - minPortID;
         private const int maxAttempts = 200;
-        private static Random RNJesus;
+        private static readonly Random _randomNumberDiety;
+        private static readonly HashSet<int> _usedPorts = new HashSet<int>();
 
         static RandomPortGenerator()
         {
             var seed = Process.GetCurrentProcess().Id + AppDomain.CurrentDomain.Id + Environment.TickCount;
-            RNJesus = new Random(seed);
+            _randomNumberDiety = new Random(seed);
         }
 
+
+        private static readonly object _lock = new object();
         public static int NextPort()
         {
-            for (var countAttempts = 0; countAttempts < maxAttempts; countAttempts++)
+            lock (_lock)
             {
-                var potentialPort = RNJesus.Next(portPoolSize) + minPortID;
-                if (IsPortAvailable(potentialPort))
+                for (var countAttempts = 0; countAttempts < maxAttempts; countAttempts++)
                 {
-                    return potentialPort;
+                    var potentialPort = _randomNumberDiety.Next(portPoolSize) + minPortID;
+                    if (!_usedPorts.Contains(potentialPort) && IsPortAvailable(potentialPort))
+                    {
+                        _usedPorts.Add(potentialPort);
```

Comment:
This is cool!  Do we want to have the test harness "release" used ports when a web test app shuts down?

Reply:
Good suggestion, I was holding off on that pending a more advanced IPC mechanism, but it was fairly easy to add here!